casper/src/main/scala/coop/rchain/casper/BlockCreator.scala

```diff
@@ -146,17 +148,19 @@ object BlockCreator {
       validDeploys = deploys.filter(
         d => notFutureDeploy(currentBlockNumber, d) && notExpiredDeploy(earliestBlockNumber, d)
       )
-      deploysInCurrentChain <- DagOperations
-                                .bfTraverseF[F, BlockMessage](parents.toList)(
-                                  b =>
-                                    ProtoUtil
-                                      .unsafeGetParentsAboveBlockNumber[F](b, earliestBlockNumber)
-                                )
-                                .map { b =>
-                                  ProtoUtil.deploys(b).flatMap(_.deploy)
-                                }
-                                .toList
-    } yield (validDeploys -- deploysInCurrentChain.flatten).toSeq
+      result <- DagOperations
+                 .bfTraverseF[F, BlockMetadata](parents.toList)(
+                   b =>
+                     ProtoUtil
+                       .unsafeGetParentMetadatasAboveBlockNumber[F](b, earliestBlockNumber, dag)
+                 )
+                 .foldLeftF(validDeploys) { (deploys, blockMetadata) =>
+                   for {
+                     block        <- ProtoUtil.unsafeGetBlock[F](blockMetadata.blockHash)
```

Comment:
I was going to say that maybe we're actually avoiding some `getBlock`s in here - but no, it seems every BlockMedatada usage added is followed by BlockMessage fetching.

Not saying that it's wrong - it's a change for the better (we restrict the scope where BlockMessage seems to be needed). It's just that it won't bring a performance imrpovement.

FYI @sebtomba

Reply:
> it seems every BlockMedatada usage added is followed by BlockMessage fetching

This is not entirely true. Previous implementation could go to BlockStore for fetching a block which was an ancestor to several initial blocks (those in `parents.toList`) multiple times while the new implementation will do this only once. LRU cache would help with this problem but since adding it is controversial (CC @sebtomba), this is an optimisation.

> It's just that it won't bring a performance imrpovement.

Performance improvement is not the only goal here. We should restrict the amount of BlockMessages simultaneously loaded into memory since they can be of unlimited size. This concern was raised by @sebtomba last Friday.

Reply:
@ArturGajowy @itegulov Yes, we are killing here two birds with one stone. Very nice!

However, I would like to go even one step further and optimize `DagOperations.bfTraverseF`. The current signature is
```scala
def bfTraverseF[F[_]: Monad, A](start: List[A])(neighbours: A => F[List[A]]): StreamT[F, A] 
```
I would like to pass the already visited neighbors hash set to the function which calculates next neighbors
```scala
def bfTraverseF[F[_]: Monad, A](start: List[A])(neighbours: (A, HashSet[A]) => F[List[A]]): StreamT[F, A]
```

In the current implementation `DagOperations.bfTraverseF` filters the neighbors **after** they have been fetched (again) from the store:
```scala
for {
  ns      <- neighbours(curr)
  visited = prevVisited + curr
  newQ    = rest.enqueue[A](ns.filterNot(visited))
} yield StreamT.cons(curr, Eval.always(build(newQ, visited)))
```

We can postpone this change for later because it requires many modifications.